src/lib/Simplify.hs

```diff
@@ -228,21 +232,43 @@ simplifyApp f xs = case fromNaryLam (NE.length xs) f of
       dropSubst $ simplifyExpr $ Case e alts' resultTy eff
     _ -> naryApp f $ toList xs
 
+-- TODO: de-dup this and simplifyApp?
+simplifyTabApp :: (Emits o, Simplifier m) => Atom o -> NonEmpty (Atom o) -> m i o (Atom o)
+simplifyTabApp f xs = case fromNaryTabLam (NE.length xs) f of
+  Just (bsCount, NaryLamExpr bs _ body) -> do
+    let (xsPref, xsRest) = NE.splitAt bsCount xs
+    ans <- dropSubst $ extendSubst (bs@@>(SubstVal <$> xsPref)) $ simplifyBlock body
+    case nonEmpty xsRest of
+      Nothing    -> return ans
+      Just rest' -> simplifyTabApp ans rest'
+  _ -> case f of
+    ACase e alts ty -> do
+      -- TODO: Don't rebuild the alts here! Factor out Case simplification
+      -- with lazy substitution and call it from here!
+      resultTy <- getTabAppType ty $ toList xs
+      alts' <- forM alts \(Abs bs a) -> do
+        buildAlt (EmptyAbs bs) \vs -> do
+          a' <- applySubst (bs@@>vs) a
+          naryTabApp a' (map sink $ toList xs)
+      dropSubst $ simplifyExpr $ Case e alts' resultTy Pure
+    _ -> naryTabApp f $ toList xs
```

Comment:
I've done a bunch of changes to `simplifyApp` that have unfortunately made it a bit bigger, so I'd really like to fold those two into the same function 😕